Tools for deterministic builds of standalone binaries and distribution archives

achow101 opened the pull request:
This PR adds several scripts and tools for making standalone binaries of the `hwi.py` script and for creating distribution archives that will go on pypi.org.

To achieve deterministic builds, the dependencies used must be locked to specific versions and hashes. To do this, I have added configuration files for using the [Poetry dependency manager](https://github.com/sdispater/poetry). Because Poetry uses a `pyproject.toml` file instead of `setup.py`, I have created a helper script which will automatically generate the proper `setup.py` file from `pyproject.toml`. The reason I chose Poetry instead of Pipenv for this task is because it has the ability to do deterministic builds of the distribution archives (python wheel and source tar for pypi.org) which Pipenv does not have.

Additionally scripts have been added to the newly created `contrib/` folder which will perform the deterministic builds of the binaries and distribution archives. The builds of the binaries are done using [pyinstaller](http://www.pyinstaller.org/). In order to build for different platforms, the `contrib/build_bin.sh` script needs to be run on each of the platforms we wish to release for. It can also be run in wine to do windows builds (see `contrib/build_wine.sh`). The configuration file that pyinstaller needs has also been added.

Lastly the pyenv version for this project has been bumped to 3.6.8 since using python 3.5.x produced standalone binaries that did not work.

This PR is built on #120 as reducing the number of dependencies fixed several issues with the standalone binary builds.

Sjors commented:
Concept ACK. Will review after 120 is merged.

instagibbs commented on docs/release-process.md:
```diff
@@ -0,0 +1,8 @@
+# Release Process
+
+1. Bump version number in `pyproject.toml`, generate the setup.py file, and tag release
```
git tag? what tag?

achow101 replied:
Git tag.

Clarified.

achow101 commented:
The best way to review this is probably to just have everyone do the build and check that their hashes match. All of the builds can be done with `contrib/build_bin.sh && contrib/build_dist.sh && contrib/build_wine.sh`. Note that `contrib/build_bin.sh` builds the binary for your type of system, so it will differ between e.g. linux and MacOS.

These are what I got:

```
$ sha256sum dist/*
e5009335aa1a73edaf209182af271c812c686aee6da6831d4ccb442bdade42a4  dist/hwi
ceae5bfc684f3f17cf25e901055893734b69a3c93ea131e1292c7eabdf6c14b4  dist/hwi-0.0.5-py3-none-any.whl
b250bff5c054ffcc08163ac42d0879b9e098b57bf287c4287ffe72118c96d88c  dist/hwi-0.0.5.tar.gz
167b96ed351bddefa00440f8f9be0453ea46c91aa7dee1e2feffbddb571ac688  dist/hwi.exe
```

Sjors commented:
I had to jump through some hoops on macOS, including `PYTHON_CONFIGURE_OPTS="--enable-framework" pyenv install 3.6.8` (see https://github.com/pyenv/pyenv/issues/1095). Hopefully that doesn't impact _using_ the binaries.

For macOS maybe point to `brew install libfaketime wine p7zip`

The wine part doesn't work for me, see [logs](https://gist.github.com/Sjors/b6e4638c4738d9975d251d8bc8570de8).

I'm get the same checksum for `dist/hwi-0.0.5-py3-none-any.whl`, but the others are different.

```
ef74d3df7a63876fa6cb5db8026294eacdb60a3909dc9b65f7c1274269ac8406  dist/hwi
ceae5bfc684f3f17cf25e901055893734b69a3c93ea131e1292c7eabdf6c14b4  dist/hwi-0.0.5-py3-none-any.whl
bdc1a36230a99a2e5c222ef16831ddef3fb9301ffc7a450a8c69eed161c3fcc6  dist/hwi-0.0.5.tar.gz
```

achow101 commented:
For macOS, I had issues getting faketime to work. I have a [fix for poetry](https://github.com/sdispater/poetry/pull/870) that removes the need to use faketime but is waiting to be merged.

fanquake commented on docs/release-process.md:
```diff
@@ -0,0 +1,43 @@
+
+Now install [Poetry](https://github.com/sdispater/poetry) with `pip install --user poetry`
+
+You will also need to install `faketime`. On Ubuntu/Debian: `sudo apt install faketime`. On Mac: `brew install faketime`
```
`brew install libfaketime`

fanquake commented:
Nice. Getting setup was easy.

Building with 88ce6ab03e6e51b33d09e8293a65bb86c9a00f70, I'm getting these hashes:
```
976354df3090a8b910fea64889481528be5db5ae75ec6b7771521aa90d7a2c8b  HWI/dist/hwi
ceae5bfc684f3f17cf25e901055893734b69a3c93ea131e1292c7eabdf6c14b4  HWI/dist/hwi-0.0.5-py3-none-any.whl
a4179a11c38cb7e6f84a1b2e2d93c912a3eb035b661a7a3172cb48316649cdbd  HWI/dist/hwi-0.0.5.tar.gz
```

When I try and do the windows build `contrib/build_wine.sh`, I'm seeing a page fault.
Backtrace and log prior:

[backtrace.txt](https://github.com/bitcoin-core/HWI/files/2917799/backtrace.txt)

```
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
003b:fixme:ntdll:NtQueryInformationJobObject stub: 0x48 9 0x22f960 144 0x22f8d0
003d:fixme:msvcrt:_configure_wide_argv (1) stub
003d:fixme:msvcrt:_initialize_wide_environment stub
003d:fixme:winsock:set_dont_fragment IP_DONTFRAGMENT for IPv6 not supported in this platform
003d:fixme:ntdll:server_get_file_info Unsupported info class e
003d:fixme:ntdll:server_get_file_info Unsupported info class e
0040:fixme:ntdll:server_get_file_info Unsupported info class e
0042:fixme:ntdll:NtQueryInformationJobObject stub: 0x20 9 0x23e420 144 0x23e390
0044:fixme:msvcrt:_configure_wide_argv (1) stub
0044:fixme:msvcrt:_initialize_wide_environment stub
0044:fixme:ntdll:server_get_file_info Unsupported info class e
0044:fixme:ntdll:server_get_file_info Unsupported info class e
0044:fixme:ntdll:server_get_file_info Unsupported info class e
wine: Unhandled page fault on read access to 0xffffffffffffffff at address 0x14003301d (thread 0042), starting debugger...
```

achow101 commented:
I've investigated the non-determinism and I think I've finally fixed it. Since for linux and macOS we are using pyenv which self compiles python, the python installation that each person was using was non-deterministic and thus introducing non-determinism. There are now instructions in `docs/release-process.md` which explain how to create a deterministic python install with pyenv. To test this, you'll need to uninstall Python 3.6.8 from pyenv first, and then you can install the deterministic one.

I've also added more to the `build_bin.sh` script which makes the python library files included in the build to be deterministic as building python deterministically is not enough.

Lastly I've added another travis job that builds the binary and checks that it works.

Below are the hashes of the result of the complete build.
```
14126ca063b72e357b36b52a3e44b413c48ed707c3ffb92af93da209feac1898  dist/hwi
7050ab9b4c50c3735d64ed02834b4528b4bf44b6c90cf5f7316f8bce045b03f2  dist/hwi-0.0.5-py3-none-any.whl
e74a8a333deb1aabcee23f17bda063edeb68de93f0bdd9ffd8c92f66817229d8  dist/hwi-0.0.5.tar.gz
2bea2fd3e3ea55ec75a28300eefafa8fd58e31b175226c330ab1fe650693b9c4  dist/hwi.exe
```

I haven't done the macOS build yet. I will do so once I get to a mac. I will also investigate the wine issue on macOS.

fanquake commented:
Tested building at 464c51db26ecced6624f7dcac86174fd64853687:

zipped `/dist` - [dist.zip](https://github.com/bitcoin-core/HWI/files/2924134/dist.zip).

```
c59c643acd2c3ed2a75ba6229674e9e24743ab8b76689baef2f283dff3d4b286 dist/hwi
7050ab9b4c50c3735d64ed02834b4528b4bf44b6c90cf5f7316f8bce045b03f2 dist/hwi-0.0.5-py3-none-any.whl
77bad3308ed5bb9a3c85cc24076abd37759b57e68a00eba50a0baa52b6646639 dist/hwi-0.0.5.tar.gz
```

`contrib/build_wine.sh` fails with:
```
Extracting archive: libusb.7z
--
Path = libusb.7z
Type = 7z
Physical Size = 980895
Headers Size = 682
Method = LZMA2:23 BCJ
Solid = +
Blocks = 2

Everything is Ok

Folders: 18
Files: 32
Size:       6294517
Compressed: 980895
0032:fixme:msvcrt:_configure_wide_argv (1) stub
0032:fixme:msvcrt:_initialize_wide_environment stub
Traceback (most recent call last):
  File "c:\python3\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\python3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "c:\python3\lib\site-packages\pip\__main__.py", line 16, in <module>
    from pip._internal import main as _main  # isort:skip # noqa
  File "c:\python3\lib\site-packages\pip\_internal\__init__.py", line 4, in <module>
    import locale
  File "c:\python3\lib\locale.py", line 16, in <module>
    import re
  File "c:\python3\lib\re.py", line 283, in <module>
    _pattern_type = type(sre_compile.compile("", 0))
  File "c:\python3\lib\sre_compile.py", line 579, in compile
    groupindex, indexgroup
TypeError: compile() argument 6 must be tuple, not list
```

achow101 commented:
It looks like faketime isn't working on mac which is why dist/hwi-0.0.5.tar.gz have different hashes.

achow101 commented:
@fanquake I got something different for the Mac build:

```
7050ab9b4c50c3735d64ed02834b4528b4bf44b6c90cf5f7316f8bce045b03f2  hwi-0.0.5-py3-none-any.whl
3083c8788120e4b4dd06a4471c29981ec690bbe6dd59ed1c7da5b3f282a94abf  hwi-0.0.5.tar.gz
53890b06a5e09c48ec0e6019b9d514dba27239df6933615befd7ab05622fa342  hwi.exe
7c6dac10288edc4623ebe31d708949d1a44f929bbaea0a05be86f3738f32d5dc  hwi
```

Can you sent me the result of
```
PYTHONHASHSEED=42 poetry run pyinstaller --additional-hooks-dir='contrib/pyinstaller-hooks/' hwi.py
```
This will create a folder named `hwi` in `dist/` instead of the executable file. It's easier to debug with that than the single executable file.

When you ran `contrib/build_bin.sh`, did you get a bunch of errors about `touch` and `strip` not recognizing arguments?

Also I had no issues with getting wine to work. I was using macOS 10.12.6 with wine installed via homebrew.

fanquake commented:
> PYTHONHASHSEED=42 poetry run pyinstaller --additional-hooks-dir='contrib/pyinstaller-hooks/' hwi.py

[dist-hwi-folder.zip](https://github.com/bitcoin-core/HWI/files/2928962/dist-hwi-folder.zip)

> When you ran contrib/build_bin.sh, did you get a bunch of errors about touch and strip not recognizing arguments?

Yep:
```
touch: illegal option -- d
usage:
touch [-A [-][[hh]mm]SS] [-acfhm] [-r file] [-t [[CC]YY]MMDDhhmm[.SS]] file ...
```

> Also I had no issues with getting wine to work on a Mac. I was using macOS 10.12.6 with wine installed via homebrew.

I'm still seeing the `TypeError: compile() argument 6 must be tuple, not list` from above. 
macOS `10.14.3`, wine also installed via brew:
```
michael$ brew info wine
wine: stable 4.0 (bottled), HEAD
Run Windows applications without a copy of Microsoft Windows
https://www.winehq.org/
/usr/local/Cellar/wine/4.0 (8,989 files, 656.4MB) *
  Poured from bottle on 2019-03-01 at 11:08:06
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/wine.rb
==> Dependencies
Build: cmake ✔, makedepend ✘, pkg-config ✔
==> Requirements
Build: macOS <= 10.12 ✘
Required: macOS >= 10.11 ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
You may also want winetricks:
  brew install winetricks
```

Do we need` winetricks` installed?

achow101 commented:
> Do we need `winetricks` installed?

No

I've just tried with a macOS 10.14.2 system and did not receive any errors with the wine build.. I honestly have no idea why it doesn't work.

I'm thinking getting deterministic builds across different machines is not going to be possible. It seems that Python itself is not deterministically built and because pyinstaller packages Python itself into the resulting binary, it will be non-deterministic so long as Python is non-deterministic. The best that I have been able to do is deterministic on the same machine.

The only one that will probably be deterministic is the Windows build as that uses Python's prebuilt binaries. I suppose the same could be done for macOS. However no such binaries exist for linux.

achow101 commented:
I've added a dockerfile and instructions to build the linux, windows, and pypi distributions using docker. This should be deterministic. Unfortunately we can't build macOS deterministically so that will remain non-deterministic.

```
$ sha256sum dist/*
22772246a469ddf5943bf2b1e8f415d754c4f1358015361dbbf425026a28555d  dist/hwi
7050ab9b4c50c3735d64ed02834b4528b4bf44b6c90cf5f7316f8bce045b03f2  dist/hwi-0.0.5-py3-none-any.whl
f450ecd14dcbb508321bc3d7151a516f94b6ba3876767a8615314d2ff6410279  dist/hwi-0.0.5.tar.gz
4e7130b4eef7a76d262a4a0a4139baf9304b10b42390b6ad89f50e9ee3cf0cbb  dist/hwi.exe
```

achow101 commented:
I've updated the dockerfile to use debian stretch-slim as the base image. There's also a travis job that runs the determinstic build, tests the linux binary, and outputs the hashes of the result. Another travis job does the macOS build but does not run any tests.

The hashes I got were:
```
503f375e671264dd6ab83f33d06568209e0a6a50773c12694f9d5b9df552383b  dist/hwi
7050ab9b4c50c3735d64ed02834b4528b4bf44b6c90cf5f7316f8bce045b03f2  dist/hwi-0.0.5-py3-none-any.whl
f450ecd14dcbb508321bc3d7151a516f94b6ba3876767a8615314d2ff6410279  dist/hwi-0.0.5.tar.gz
4e7130b4eef7a76d262a4a0a4139baf9304b10b42390b6ad89f50e9ee3cf0cbb  dist/hwi.exe
```

fanquake commented:
I'm seeing the same hashes as you:
```
root@e10a4cdc76f4:/opt/hwi# sha256sum dist/*
503f375e671264dd6ab83f33d06568209e0a6a50773c12694f9d5b9df552383b  dist/hwi
7050ab9b4c50c3735d64ed02834b4528b4bf44b6c90cf5f7316f8bce045b03f2  dist/hwi-0.0.5-py3-none-any.whl
f450ecd14dcbb508321bc3d7151a516f94b6ba3876767a8615314d2ff6410279  dist/hwi-0.0.5.tar.gz
4e7130b4eef7a76d262a4a0a4139baf9304b10b42390b6ad89f50e9ee3cf0cbb  dist/hwi.exe
```

As mentioned, sometimes the Windows build hangs (using `contrib/build_wine.sh`), particularly when installing dependencies via `poetry`. However running the commands manually inside the container works fine..

achow101 commented:
I've made a few changes to `.travis.yml` to build the macOS binary with an older version for better compatibility. That travis job will also upload the resulting binary to an AWS S3 bucket. That upload will be what I will use for releases.

Can someone with a mac test that this binary works and actually communicates with devices? It can be downloaded from https://s3.amazonaws.com/hwi-travis-builds/achow101/hardware-wallet-interface/114/114.4/hwi. This was built in a travis job on my personal fork of this repo: https://travis-ci.com/achow101/hardware-wallet-interface/jobs/183298042.

Sjors commented:
```sh
wget https://s3.amazonaws.com/hwi-travis-builds/achow101/hardware-wallet-interface/114/114.4/hwi
sha256sum hwi
d9acc15b507ce2079134b372e68278f71780ff0c762249c906778aefdf47f5f4  hwi
```

`enumerate` and `getkeypool` work

achow101 commented:
Merging as there is agreement on the produced binary hashes and that they all work.
